### PR TITLE
Specify the OpenTelemetry Collector config as YAML

### PR DESCRIPTION
**Pull request by iblancasa**
Relates to #1707.

This PR introduces the following changes:
* Add the `.spec.configSpec` field to the OpenTelemetry Collector CRD

TODO:
* Fix the unit tests
* Fix the E2E tests
* Add an E2E test to ensure the old behavior continues working
* Mark the `.spec.config` field as deprecated
* Use the new structure across all the operator

### CODE REVIEW

**Comment by iblancasa** [comment]
Some notes:
* I tried the different approaches described in the different edits from [this comment](https://github.com/open-telemetry/opentelemetry-operator/issues/1707#issuecomment-1721231339) but `controller-gen` always complains
* I tried adding a field to the `OpenTelemetryCollector` type (it would be the ideal thing) in order to store the representation of the information we want. I end up adding it to the `Params` structure (maybe not the best option)
* Using `runtime.RawExtension` following some recommendations from people in the `kubebuilder` channel from the Kubernetes Slack workspace for similar situations

**Comment by pavolloffay** [comment]
We have discussed this at the SIG meeting. We might want to bump the CRD API version and just use the typed config

**Comment by iblancasa** [comment]
> We have discussed this at the SIG meeting. We might want to bump the CRD API version and just use the typed config

The meetings are no longer recorded, right? Can you, at least, add something to #1707?

File: apis/v1alpha1/config.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"encoding/json"
+
+	"gopkg.in/yaml.v2"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func MustParseConfigSpec(in string) *ConfigSpec {
+	cfg := &ConfigSpec{}
+	yaml.Unmarshal([]byte(in), &cfg)
```
Comment (frzifus) [inline apis/v1alpha1/config.go:12]:
```suggestion
	if err := yaml.Unmarshal([]byte(in), &cfg); err != nil {
		panic(err)
	}
```

File: internal/manifests/params.go [hunk 2]
```diff
@@ -24,8 +28,75 @@ import (
...
+	configStr string
+	configSpec v1alpha1.ConfigSpec
+}
+
+func (c *OtelConfig) String() string {
+	if c.configStr != "" {
+		return c.configStr
+	}
+	return c.configSpec.String()
+}
```
Comment (frzifus) [inline internal/manifests/params.go:95]: Why is this otel config needed? Doesn't it duplicate the data from the OpenTelemetryCollector CR?

File: internal/manifests/params.go [hunk 3]
```diff
@@ -24,8 +28,75 @@ import (
...
 
+func NewOtelConfig(instance *v1alpha1.OpenTelemetryCollector) (OtelConfig, error){
+	var cfg OtelConfig
+	if instance.Spec.Config != "" {
+		cfg.configStr = instance.Spec.Config
+		err:=yaml.Unmarshal([]byte(instance.Spec.Config), &cfg)
+		if err!= nil {
+			return cfg, fmt.Errorf("there was a problem parsing the config: %w", err)
+		}
+	} else {
```
Comment (frzifus) [inline internal/manifests/params.go:39]: Does it make things simpler when we use an early return instead?

File: controllers/opentelemetrycollector_controller.go [hunk 4]
```diff
@@ -175,7 +175,13 @@ func isNamespaceScoped(obj client.Object) bool {
 }
 
 func (r *OpenTelemetryCollectorReconciler) getParams(instance v1alpha1.OpenTelemetryCollector) manifests.Params {
+	otelCfg, err:= manifests.NewOtelConfig(&instance)
+	if err!= nil {
+		r.log.Error(err, "there was a problem while parsing the config")
+	}
+
```
Comment (frzifus) [inline controllers/opentelemetrycollector_controller.go:182]: Should we somehow cancel the reconciliation in case of an error?

**Review by frzifus — COMMENTED** [review]
> I tried the different approaches described in the different edits from https://github.com/open-telemetry/opentelemetry-operator/issues/1707#issuecomment-1721231339 but controller-gen always complains

What did it complain about?

>  I tried adding a field to the OpenTelemetryCollector type (it would be the ideal thing) in order to store the representation of the information we want. I end up adding it to the Params structure (maybe not the best option)

Why is that needed?

> > We have discussed this at the SIG meeting. We might want to bump the CRD API version and just use the typed config

> The meetings are no longer recorded, right? Can you, at least, add something to https://github.com/open-telemetry/opentelemetry-operator/issues/1707?

@iblancasa you can find the recording on [this list](https://docs.google.com/spreadsheets/d/1SYKfjYhZdm2Wh2Cl6KVQalKg_m4NhTPZqq-8SzEVO6s). Here are the [meeting notes](https://docs.google.com/document/d/1Unbs2qp_j5kp8FfL_lRH-ld7i5EOQpsq0I4djkOOSL4/edit#heading=h.okck2w75mwlk).

My internet connection was not so stable in the last meeting. But from what I have gathered, the idea is to raise api version for breaking change of config field and [provide automatic CR  conversion](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#webhook-conversion).

**Pull request closed without merging**